src/sync/push_client.cpp

```diff
@@ -0,0 +1,90 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/push_client.hpp"
+#include "sync/app_utils.hpp"
+
+namespace realm {
+namespace app {
+
+void PushClient::register_device(const std::string& registration_token,
+                                 std::shared_ptr<SyncUser> sync_user,
+                                 std::function<void(util::Optional<AppError>)> completion_block)
+{
+    auto handler = [completion_block](const Response& response) {
+        if (auto error = check_for_errors(response)) {
+            return completion_block(error);
+        } else {
+            return completion_block({});
+        }
+    };
+    auto push_route = util::format("/app/%1/push/providers/%2/registration", m_app_id, m_service_name);
+    std::string route = m_auth_request_client->url_for_path(push_route);
+    
+    bson::BsonDocument args {
+        {"registrationToken", registration_token}
+    };
+    
+    std::stringstream s;
+    s << bson::Bson(args);
+    
+    Request request {
+        .method = HttpMethod::put,
```

Comment:
Looks like there's a problem using this syntax in vs2019 (https://github.com/realm/realm-object-store/pull/1007). Probably need to initialize members explicitly, or maybe it's time we make a constructor for this struct?

Reply:
Would it be a faux-pas to make a constructor for a struct that is effectively a data bag?

Reply:
Maybe it's a bit heavy handed for this type of struct, but it would be clear. I think the problem was that this is a C++ 20 feature that isn't implemented in VS2019 yet, but we should be able to use the unnamed version since it should be supported since C++11 (https://en.cppreference.com/w/cpp/language/aggregate_initialization) so an alternative to making a constructor that should also work in VS could be `Request request { HttpMethod::put, route, m_request_timeout_ms, {}, s.str() }`

Reply:
changed to use:

`m_auth_request_client->do_authenticated_request({
        HttpMethod::del,
        route,
        m_timeout_ms,
        {},
        s.str(),
        false
    },
    sync_user,
    handler);`